### PR TITLE
track prints to GA

### PR DESCRIPTION
**Pull request by jpslav**
Tracks prints to Google Analytics.  The event is registered after the user finishes or cancels the print action.

image

### CODE REVIEW

**Review by TomWoodward — CHANGES_REQUESTED** [review]

File: src/app/content/print.ts [hunk 1]
```diff
@@ -1,20 +1,11 @@
...
-export const print = () => {
+export const print = (book: Book | undefined, currentPath: string) => {
   const window: Window = assertWindow();
   window.print();
 
-  const pathname: string = window.location.pathname;
-
-  if (pathname) {
-    let slug: string = 'unknown';
-    const slugMatches: string[] | null = pathname.match(/\/books\/(.*)\/pages*/);
```
Comment (jpslav) [inline src/app/content/print.ts]: goodbye sweet regex designed to annoy tom

File: src/app/content/print.spec.ts [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+    originalLocation = window.location;
+    delete window.location;
+
+    client.setTrackingIds(['foo']);
+    jest.resetAllMocks();
+  });
+
+  it('prints', async() => {
+    window.location = {...originalLocation, pathname};
```
Comment (TomWoodward) [inline src/app/content/print.spec.ts]: this isn't necessary here right?

File: src/app/content/print.spec.ts [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    await print(book, pathname);
+    expect(mockPrint).toHaveBeenCalled();
+  });
+
+  it('with good pathname tracks to GA', async() => {
+    window.location = {...originalLocation, pathname};
```
Comment (TomWoodward) [inline src/app/content/print.spec.ts]: or this?

File: src/app/content/print.spec.ts [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    await print(book, pathname);
+    expect(mockGa).toHaveBeenCalledWith('tfoo.send', 'event', 'REX print', 'biology-2e', pathname);
+  });
+
+  it('with undefined pathname tracks to GA', async() => {
+    window.location = {...originalLocation, pathname: undefined};
```
Comment (TomWoodward) [inline src/app/content/print.spec.ts]: or this?

File: src/app/content/print.spec.ts [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    window.print = mockPrint;
+    mockGa = jest.fn<UniversalAnalytics.ga, []>();
+    window.ga = mockGa;
+
+    originalLocation = window.location;
+    delete window.location;
+
+    client.setTrackingIds(['foo']);
+    jest.resetAllMocks();
+  });
```
Comment (TomWoodward) [inline src/app/content/print.spec.ts]: this tests leaks globals for `window.ga` and `window.print` you should back up the original values and restore them in an `afterEach` like this:

```js
const printBack = window.print;
beforeEach(() => {
  window.print = mockPrint;
});
afterEach(() => {
  window.print = printBack;
})
```

and for the ga too

**Review by TomWoodward — CHANGES_REQUESTED** [review]

**Review by TomWoodward — APPROVED** [review]

**Pull request merged**